### ISSUE TITLE
File Nesting Not Working for all Class or Shared Library Projects

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/project-system#5722]
_This issue has a corresponding [ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/483587/file-nesting-not-working-for-all-class-or-shared-l.html). Please vote and comment there to make sure your voice is heard._

_Copied to [AB#986508](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/986508)_

---
File Nesting is just not working for all Class or Shared Library Projects

---
#### Visual Studio Feedback System on 11/03/2019, 05:51 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Drew Noakes [MSFT] on 13/03/2019, 00:19 AM: 

<p>Thanks for your feedback. Could you please provide some more details to help us investigate this issue? A screenshot would be great, along with what files you expect to appear nested. Also please confirm whether you're seeing the nesting you expect in certain project types, and if so, which ones.</p>

#### Kearin on 08/05/2019, 00:12 AM: 

This might be when you create a class A.cs and then auto generate code into a file A.B.cs. It used to be that A.B.cs would nest under A.cs in the solution explorer. This behavior was similar to specifying A.cs in the project file.<br>
#### Alex Engblom on 25/06/2019, 05:37 PM: 

<p>At least the basic 'extensionToExtension' rule of nesting .cs.d.ts -&gt; .cs does not seem to work in class libraries. This would be great to trim out the common entities project etc... Doesn't work with changing Standard Settings to 'Web' nor by creating custom solution level .filenesting.json file with appropriate settings. Tested with VS2017 and VS2019.</p>

#### David Adams on 04/07/2019, 10:21 AM: 

<p>This is so frustrating because file nesting is essential to any large project.  It was so simple with non .NET core projects with Mads K. extension because this was completely hands-off and worked automatically.  It does not appear, nor is it made clear, that the filenesting.json works with class libraries.  The workaround seems to be to edit each and every csproj file for class libraries and individually hard-code each nested file such as:</p><pre> &lt;Compile Update="$(ProjectDir)\Org.*.cs"&gt;
      &lt;DependentUpon&gt;$(ProjectDir)\Org.cs&lt;/DependentUpon&gt;
    &lt;/Compile&gt;</pre><p class="">However, file nesting is lost when the project is unloaded for nested files which are in subfolders.  So the following:</p><pre class=""> &lt;Compile Update="$(ProjectDir)\Model\Org.*.cs"&gt;
      &lt;DependentUpon&gt;$(ProjectDir)\Model\Org.cs&lt;/DependentUpon&gt;
    &lt;/Compile&gt;<br></pre><p class="">will work temporarily but when you exit VS and reload the file nesting will be lost for nested files in subfolders.  The only way to get it back is to re-edit the csproj file and modify each dependency again to trigger the file nesting.   It would really be nice if the same filenesting settings for a VS .NET core solution would apply to all projects and not just the web project.  I am using VS 2019 16.1.4.</p>
</pre>

#### Alex Engblom on 04/07/2019, 07:38 PM: 

<p>Just found out this issue in github https://github.com/dotnet/project-system/issues/3242 . Workaround suggested by <a rel="user" href="https://developercommunity.visualstudio.com/users/94131/d52c28a6-1cf4-6ced-bd34-e6dec6e8d18f.html" nodeid="94131">@Drew Noakes [MSFT]</a> - adding following elements in .csproj did the trick for me. For now. For .NET Standard 2.0, didn't work with Framework 4.7.1.</p><p>&lt;ItemGroup&gt;     <br>&lt;ProjectCapability Include="DynamicDependentFile" /&gt;     <br>&lt;ProjectCapability Include="DynamicFileNesting" /&gt; <br>&lt;/ItemGroup&gt;<br></p>

#### David Adams on 05/07/2019, 09:24 AM: 

<p>Nice find Alex E - I will follow that github thread.  Unfortunately the addition of the ProjectCapability elements to the .csproj file this does not seem to help with .NET core class libraries which have file nesting in project subfolders.  Just as before, once the project is reloaded the file nesting is lost in any subfolders, and the only workaround is to re-edit the .csproj file and save while the project is open.</p>

#### Visual Studio Feedback System on 25/07/2019, 08:51 AM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

**Comment by NickCraver** [comment]
FWIW, we are doing the `Directory.Build.props` approach as well:
```xml
<ItemGroup>
  <ProjectCapability Include="DynamicDependentFile" />
  <ProjectCapability Include="DynamicFileNesting" />
</ItemGroup>
```

It it helps, the reason I was given last time on this being restricted to ASP.NET Web projects by default is it's supported by _that workload in Visual Studio_. So the assumption is: dealing with ASP.NET? You're on that workload...or things are funky anyway. So that's where it's lit up.

To light it up for all projects, it'd have to move to another workload in Visual Studio, which means a team ownership change...and I'm not sure how that part works :)

**Comment by drewnoakes** [comment]
@NickCraver we are investigating the cross-team move, to make the configurable nesting implementation that's currently available in web projects available for all SDK-style projects. This wouldn't require any special workloads beyond those you need to open SDK-style projects anyway.

**Comment by swesonga** [comment]
A fix for this issue is in Visual Studio 16.7 Preview 1 and will need to be enabled by checking the Tools > Options > Environment > Preview Features > **Enable Configurable File Nesting** option then restarting Visual Studio.

**Comment by BrunoBlanes** [comment]
Lovely fix.

**Comment by anrichter** [comment]
Hi,

I miss this feature in the final version from Visual Studio 16.7. (16.7.1)
The Visual Studio 16.8 Preview contains the file nesting feature.

**Comment by swesonga** [comment]
> Hi,
> 
> I miss this feature in the final version from Visual Studio 16.7. (16.7.1)
> The Visual Studio 16.8 Preview contains the file nesting feature.

Sorry there was no additional follow-up here. This feature is still only usable in the Preview builds because #1176 needs to be fixed before enabling it in the RTM channel.

**Comment by anrichter** [comment]
Thanks for the explanation.

**Comment by julealgon** [comment]
@swesonga could you please elaborate why exactly fixing #1176 is required to promote this to RTM?

Would it be possible to untie these issues from one another? #1176 has been open for 3+ years now with no fix in sight. It was recently removed from 16.8 timeline as well, which means we'll probably end up another entire major VS release at least without proper file nesting support.

**Comment by swesonga** [comment]
@julealgon The reason this depends on #1176 is that the bug there can potentially start happening in more scenarios with file nesting enabled. @lifengl has a fix for that issue is in the works for 16.9.

**Comment by julealgon** [comment]
I assume it happens on more specialized scenarios right @swesonga ? Can't this be shipped as a preview feature in VS 16.8 RTM and be promoted to non-preview later in 16.9?

I guess the benefit of doing that depends on how long it will take for 16.9 to be available and the fix to be in place. Do you have any ETA?

**Comment by swesonga** [comment]
@julealgon The nesting feature allows for custom nesting rules, which can possibly introduce cycles in the dependencies. It may make sense to view that as a more specialized scenario, but the possibility of VS crashing in that scenario means the feature is not allowed to be promoted to RTM. As far as ETA, we're most likely looking at 16.10 for turning on this feature by default since 16.9 is for the loop dependencies fix.

**Comment by julealgon** [comment]
I appreciate the clarification @swesonga . I was able to get something going on my side by adding the csproj properties and using a standard Web template as a solution-level nesting configuration. This is good enough for me for now.

**Comment by swesonga** [comment]
This file nesting feature will not be turned on by default in VS versions 16.* because of performance issues to be investigated in #7161

**Comment by Metrowerks** [comment]
I have VS 16.9.2 and the file nesting still doesn't work. It only work once after the first saving of the csproj file. After that I have to make some changes and save the csproj again for the nesting to work.

**Comment by DanielD** [comment]
I am running VS 16.11.1 and this problem still exists.  Is there a special set of options to get this to work on class libraries?  Even with a performance hit, this option needs to be available to developers, regardless of the version installed.

**Comment by nathan-alden-sr** [comment]
It appears this also doesn't work in VS 2022 RC3. I have a `net6.0` class library project with the special items:
```xml
<ItemGroup>
  <ProjectCapability Include="DynamicDependentFile" />
  <ProjectCapability Include="DynamicFileNesting" />
</ItemGroup>
```
as well as a `.filenesting.json` file next to the solution with contents set to this:
```json
{
  "help": "https://go.microsoft.com/fwlink/?linkid=866610",
  "root": false,

  "dependentFileProviders": {
    "add": {
      "fileSuffixToExtension": {
        "add": {
          ".Records.cs": [
            ".cs"
          ]
        }
      }
    }
  }
}
```
No 🎲🎲.

**Comment by IanKemp** [comment]
@nathan-alden-sr Seems like this feature is turned off in all VS versions due to the aforementioned nebulous "performance issues" bug #7161. And if you look at said bug, you'll see that it's nowhere. So I'm guessing that as per standard Microsoft operating procedures, this feature will never be enabled.

**Comment by MithrilMan** [comment]
This is really annoying, I've big solutions where I had to enable file nesting similar to @nathan-alden-sr snippet and now with Visual Studio 2022 I can't have nested files anymore.

Please let us have an opt-in way to enable custom file nesting on every solution type, thanks.

**Comment by jeffward01** [comment]
> It appears this also doesn't work in VS 2022 RC3. I have a `net6.0` class library project with the special items:
> 
> ```
> <ItemGroup>
>   <ProjectCapability Include="DynamicDependentFile" />
>   <ProjectCapability Include="DynamicFileNesting" />
> </ItemGroup>
> ```
> 
> as well as a `.filenesting.json` file next to the solution with contents set to this:
> 
> ```json
> {
>   "help": "https://go.microsoft.com/fwlink/?linkid=866610",
>   "root": false,
> 
>   "dependentFileProviders": {
>     "add": {
>       "fileSuffixToExtension": {
>         "add": {
>           ".Records.cs": [
>             ".cs"
>           ]
>         }
>       }
>     }
>   }
> }
> ```
> 
> No 🎲🎲.


I am posting reply to state that this does not work for me either.  

I added the XML elements to the .csproj file as below:
```xml
<ItemGroup>
  <ProjectCapability Include="DynamicDependentFile" />
  <ProjectCapability Include="DynamicFileNesting" />
</ItemGroup>
```

**When I reload the project I receive this error:** 

![error_nesting](https://user-images.githubusercontent.com/12589359/145698324-3ff916f8-dd30-4d77-96ff-a1ef8f5e339f.png)

---
**Details:**
net 6.0 projects 
Microsoft Visual Studio Community 2022 (64-bit) 
Version 17.0.2

----

Please give us some way to say 'yes I will take the perfomance hit in bug #7161 and configure this by config file of preferably `.csproj`

**Comment by XtremeOwnageDotCom** [comment]
Bump..... Can confirm, this is STILL an issue with the latest VS2022 beta.

**Comment by fschmied** [comment]
New issue for VS 2022: #7939

@nathan-alden-sr @IanKemp @MithrilMan @jeffward01 @XtremeOwnageDotCom

**Comment by drewnoakes** [comment]
This is being actively worked on, though being tracked in #7161. There are performance implications for us to turn it on by default for all projects.

To enable configurable file nesting for non-web projects in VS 2022, add this to your `.csproj` file:

```xml
<ItemGroup>
  <ProjectCapability Include="ConfigurableFileNesting" />
  <ProjectCapability Include="ConfigurableFileNestingFeatureEnabled" />
</ItemGroup>
```

**Comment by MithrilMan** [comment]
> This is being actively worked on, though being tracked in #7161. There are performance implications for us to turn it on by default for all projects.
> 
> To enable configurable file nesting for non-web projects in VS 2022, add this to your `.csproj` file:
> 
> ```
> <ItemGroup>
>   <ProjectCapability Include="ConfigurableFileNesting" />
>   <ProjectCapability Include="ConfigurableFileNestingFeatureEnabled" />
> </ItemGroup>
> ```

Oh... so the reason why my project doesn't work is because you changed these 2 capabilities?
I was using 
```
<ItemGroup>
   <ProjectCapability Include="DynamicDependentFile" />
   <ProjectCapability Include="DynamicFileNesting" />
</ItemGroup>
```
I couldn't find any information regarding ConfigurableFileNesting and ConfigurableFileNestingFeatureEnabled, when was it introduced?
It need to be included in file nesting documentation, thanks.

**Comment by drewnoakes** [comment]
@MithrilMan they were changed in 17.0. We won't document them as project capabilities are generally internal, and the intention here is that nesting be enabled in a more natural and discoverable way. Again, this is being worked on currently.

**Comment by JacqueSatan** [comment]
> This is being actively worked on, though being tracked in #7161. There are performance implications for us to turn it on by default for all projects.
> 
> To enable configurable file nesting for non-web projects in VS 2022, add this to your `.csproj` file:
> 
> ```
> <ItemGroup>
>   <ProjectCapability Include="ConfigurableFileNesting" />
>   <ProjectCapability Include="ConfigurableFileNestingFeatureEnabled" />
> </ItemGroup>
> ```

What about C++ projects?
